Add self-hosting documentation (#419)
game-ci/documentation: docs/12-self-hosting/02-ubuntu-setup.mdx

    sudo useradd -s /bin/bash -d /home/$NEW_USER/ -m -G sudo $NEW_USER
    ```

  - Grant passwordless sudo permission

    ```bash
    sudo echo "$NEW_USER ALL=(ALL) NOPASSWD: ALL" >> /etc/sudoers
    ```

  - Import an ssh key

    ```bash
    sudo -u friend ssh-import-id-gh cloudymax
    ```

  - Add the user to relevant groups

    ```bash
    sudo usermod -a -G kvm $NEW_USER
    sudo usermod -a -G docker $NEW_USER
    ```

  - Craete a password for the user

    ```bash
    sudo passwd $NEW_USER
    ```

## Docker Compose

The default Ubuntu package lists provide a very outdated version of docker compose. Below are the
steps required to install a current version.

- Find the latest version by visiting https://github.com/docker/compose/releases

  ```bash
  export COMPOSE_VERSION="2.17.3/"
  ```

- Create a directory for the binary

  ```bash
  mkdir -p ~/.docker/cli-plugins/
  ```

- Download the binary

  ```bash
  curl -SL https://github.com/docker/compose/releases/download/v$COMPOSE_VERSION/docker-compose-linux-x86_64 -o ~/.docker/cli-plugins/docker-compose
  ```

- Make it executable

  ```bash
  chmod +x ~/.docker/cli-plugins/docker-compose
  ```

> ## NVIDIA GPU Drivers

Reviewer: ```suggestion
## Nvidia GPU Drivers
```

> Ubuntu's built-in driver installation tool is really unreliable. Instructions for its use are
> included, but readers are advised to get their driver installaer directly from NVIDIA

Reviewer: ```suggestion
You are advised to get the driver installer directly from Nvidia.
```

  make \
  pkg-config \
  libvulkan1
  ```

- Find your driver version and download the installer

  - Use Nvidia's web tool located here: https://www.nvidia.com/download/index.aspx.

  - Alternatively, you can download a specific driver version using curl

    ```bash
    export DRIVER_VERSION=""

    # GeForce Cards
    curl --progress-bar -fL -O "https://us.download.nvidia.com/XFree86/Linux-x86_64/$DRIVER_VERSION/NVIDIA-Linux-x86_64-$DRIVER_VERSION.run"

    # Datacenter Cards
    curl --progress-bar -fL -O "https://us.download.nvidia.com/tesla/$DRIVER_VERSION/NVIDIA-Linux-x86_64-$DRIVER_VERSION.run"
    ```

- Run the Installer

  - It is required to run the driver from the system console, it cannot be installed from an
    X-session.

    ```bash
    sudo bash "NVIDIA-Linux-x86_64-*.run"
    ```

### Install from apt

Automatic install is currently broken for multiple card-types, see
[#1993019](https://bugs.launchpad.net/ubuntu/+source/ubuntu-drivers-common/+bug/1993019).

- Automatic install (Broken)

  ```bash
  sudo ubuntu-drivers autoinstall
  ```

- Install specific driver version

  ```bash
  sudo ubuntu-drivers install nvidia:525
  ```

## Nvidia Container Toolkit

> The nvidia container toolkit allows containes to access the GPU resources of the underlying host
> machine. Requires that the GPU drivers are already installed on the host. See the official docs
> here: https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/install-guide.html

Reviewer: ```suggestion
The Nvidia Container Toolkit allows containers to access the GPU resources of the underlying Host. Requires that the GPU drivers are already installed on that Host. See the [official docs](https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/install-guide.html)
```

- Get your system distribution name

  ```bash
  distribution=$(. /etc/os-release;echo $ID$VERSION_ID)
  ```

- Download the gpg key and add the repo to your apt sources

  ```bash
  curl -fsSL https://nvidia.github.io/libnvidia-container/gpgkey | sudo gpg --dearmor -o /usr/share/keyrings/nvidia-container-toolkit-keyring.gpg \
  && curl -s -L https://nvidia.github.io/libnvidia-container/$distribution/libnvidia-container.list | \
      sed 's#deb https://#deb [signed-by=/usr/share/keyrings/nvidia-container-toolkit-keyring.gpg] https://#g' | \
      sudo tee /etc/apt/sources.list.d/nvidia-container-toolkit.list
  ```

- Update apt packages and install the container toolkit

  ```bash
  sudo apt-get update && sudo apt-get install -y nvidia-container-toolkit
  ```

- Set `nvidia` the default container runtime

  ```bash
  sudo nvidia-ctk runtime configure --runtime=docker --set-as-default
  ```

- Restart the docker service

  ```bash
  sudo systemctl restart docker
  ```

> - Test that it is working

Reviewer: ```suggestion
- Verify that it works
```

  > |=========================================+======================+======================|
  > |   0  Tesla M40 24GB                 On  | 00000000:04:00.0 Off |                    0 |
  > | N/A   36C    P8              28W / 250W |     23MiB / 23040MiB |      0%      Default |
  > |                                         |                      |                  N/A |
  > +-----------------------------------------+----------------------+----------------------+
  >
  > +---------------------------------------------------------------------------------------+
  > | Processes:                                                                            |
  > |  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
  > |        ID   ID                                                             Usage      |
  > |=======================================================================================|
  > |  No running processes found                                                           |
  > +---------------------------------------------------------------------------------------+

>   > ```

Reviewer: Suggest also a `:::note :::` admonition here

  ```

>   ```

Reviewer: Unused codeblock

Author: fixed!
